nats.go

```diff
@@ -2244,10 +2249,30 @@ func (nc *Conn) processPermissionsViolation(err string) {
 // processAuthorizationViolation is called when the server signals a user
 // authorization violation.
 func (nc *Conn) processAuthorizationViolation(err string) {
+	nc.processAuthError(ErrAuthorization)
+}
+
+// processAuthenticationExpired is called when the server signals a user
+// authorization has expired.
+func (nc *Conn) processAuthenticationExpired(err string) {
+	nc.processAuthError(ErrAuthExpired)
+}
+
+// processAuthError generally processing for auth errors. We want to do retries
+// unless we get the same error again. This allows us for instance to swap credentials
+// and have the app reconnect, but if nothing is changing we should bail.
+func (nc *Conn) processAuthError(err error) {
 	nc.mu.Lock()
-	nc.err = ErrAuthorization
+	nc.err = err
 	if nc.Opts.AsyncErrorCB != nil {
-		nc.ach.push(func() { nc.Opts.AsyncErrorCB(nc, nil, ErrAuthorization) })
+		nc.ach.push(func() { nc.Opts.AsyncErrorCB(nc, nil, err) })
+	}
+	// We should give up if we tried twice on this server and got the
+	// same error.
+	if nc.current.lastErr == err {
+		defer nc.Close()
```

Comment:
Hey @kozlovic should we be shutting down here or just removing this one from the list? If there are more servers in the list we probably should not give up.

Reply:
Why not.

Reply:
If I am connected to a cluster, and one server is bad but others are ok, we should be able to reconnect.

Reply:
I meant yes, why not do that. But I can't really recall why we did this in the first place. What problem exactly were we trying to solve? Is there a situation where a server will return auth error always and not other servers in the cluster? (without being a misconfig). If auth is updated in a server and a client connecting to the server while auth is returning error, we possibly can remove it from the list of urls while the server is then being updated and would have accepted that connection later on.

Reply:
If I revoke your credentials in some form or fashion I wanted the default behavior to not keep trying forever.